### PR TITLE
Json rpc API

### PR DESCRIPTION
**Pull request by xterr**
The new JSON-RPC API implements all the current methods (in different forms), plus new methods like "GetBlocksByNumbers", "GetBlocksByRange".

=== TESTS ===
Added tests for the core JsonRpc code, and some methods (more to come).
Tests can be run using: "npm run json_rpc_test"

For now, these are separated from the original tests because they need different configurations.

=== ESLINT ===
Added also eslint, with some basic, common, rules. A pre-commit git hook can be added to run and enforce the checks before a commit.

### CODE REVIEW

File: src/node/jsonRpc/Methods/SendTransaction.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        if (typeof fee !== 'undefined' && (isNumber(fee) === false || fee <= 0))
+        {
+            throw new Error('Fee is invalid. Only numerical values greater than 0 are supported.');
+        }
+
+        try
+        {
+            const aResponse = await this._oTransactionsManager.wizard.createTransactionSimple( fromAddress, toAddress, value, fee, undefined, password);
```
Comment (xterr) [inline src/node/jsonRpc/Methods/SendTransaction.js]: This works, as in the transaction is added successfully in the Pending queue, the amount is substracted from the wallet`s balance, but the transaction is not ever processed.

Need feedback on this.

File: src/node/jsonRpc/tests/Utils/BlockFinder.test.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
+import { expect, assert } from 'chai';
+
+import BlockFinder from '../../Utils/BlockFinder';
+
+describe('BlockFinderTest', () => {
+    const Blockchain = {
```
Comment (xterr) [inline src/node/jsonRpc/tests/Utils/BlockFinder.test.js]: This object can be a mock of the actual Blockchain class also, but this way is cleaner

File: tests/webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+            'node': path.resolve(__dirname + '/../src/node'),
+            'consts': path.resolve(__dirname + '/../src/consts'),
+            'common': path.resolve(__dirname + '/../src/common'),
+            'tests': path.resolve(__dirname + '/../src/tests'),
+        }
+    },
+    target: 'node', // webpack should compile node compatible code
+    externals: [nodeExternals()], // in order to ignore all modules in node_modules folder
+    devtool: "inline-cheap-module-source-map"
+};
```
Comment (xterr) [inline tests/webpack.config.js]: This is a clean webpack file, needed only for mocha tests

File: src/build_terminal.js [hunk 4]
```diff
@@ -17,6 +17,33 @@ Blockchain.createBlockchain("full-node", ()=>{}, async ()=>{
...
+
+        // @see express-rate-limit package for configuration (except isEnabled)
+        rateLimit: {
+            windowMs: 60 * 1000, // 1 minute
+            max     : 60,        // limit each IP to 100 requests per windowMs,
+            isEnabled: true
+        }
+    };
+
+    if (typeof process.env.JSON_RPC_USERNAME !== "undefined" && typeof process.env.JSON_RPC_PASSWORD !== "undefined")
```
Comment (xterr) [inline src/build_terminal.js]: If the username/password are not set, the authentication is disabled. Some methods that require authentication will not be available.

File: src/node/jsonRpc/index.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import * as oMethods from './Methods';
+
+const oLogger           = new Logger('JSON RPC Server');
+const oRpcMethodManager = new RpcMethodManager;
+oRpcMethodManager.addMethods(Object.values(oMethods));
+
+const JsonRpcServer = (oConfig) => {
+
+    // If the "serverConfig" property or the port is not defined, don`t start the JSON RPC Server
+    if (typeof oConfig['serverConfig'] === 'undefined' || oConfig['serverConfig'].port === null)
```
Comment (xterr) [inline src/node/jsonRpc/index.js]: If the serverConfig or the port is not defined, the server won`t start.

File: src/node/jsonRpc/index.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    app.use(jsonParser());
+    app.use(fRpcServerMiddleware(oRpcMethodManager, oConfig.basicAuth.isEnabled));
+
+    if (typeof oConfig['serverConfig'] === 'undefined')
+    {
+        return;
+    }
+
+    const oServerConfig = defaults(omitBy(oConfig['serverConfig'], isUndefined), {
+        host: '127.0.0.1' // By default listen only on localhost
```
Comment (xterr) [inline src/node/jsonRpc/index.js]: For added security, the server should listen, by default, only on localhost.

**Comment by morion4000** [comment]
> Added also eslint, with some basic, common, rules. A pre-commit git hook can be added to run and enforce the checks before a commit.

How can we run eslint checks?

**Comment by xterr** [comment]
My pre-commit file:
It will run eslint only for staged files + run tests (for now, only json_rpc_tests):

https://gist.github.com/xterr/c97be445e64b6b02cdc917dfd29dc362

File: src/node/jsonRpc/Methods/SendTransaction.js (hunk 1 above)
Comment (morion4000) [reply src/node/jsonRpc/Methods/SendTransaction.js]: The legacy API uses the same method to create transactions - https://github.com/WebDollar/Node-WebDollar/blob/master/src/node/sockets/node-server/API/Node-API-Private.js#L74

That might be an issue with the transaction propagation or node setup.

**Pull request closed without merging**